File: lib/src/compiler/parse_util.dart
library angular2.src.compiler.parse_util;

class ParseLocation {
  ParseSourceFile file;
  num offset;
  num line;
  num col;
  ParseLocation(this.file, this.offset, this.line, this.col) {}
  String toString() {
    return '''${ this . file . url}@${ this . line}:${ this . col}''';
  }
}

class ParseSourceFile {
  String content;
  String url;
  ParseSourceFile(this.content, this.url) {}
}

abstract class ParseError {
  ParseLocation location;
  String msg;
  ParseError(this.location, this.msg) {}
  String toString() {
    var source = this.location.file.content;
    var ctxStart = this.location.offset;
    if (ctxStart > source.length - 1) {
      ctxStart = source.length - 1;
    }
    var ctxEnd = ctxStart;
    var ctxLen = 0;
    var ctxLines = 0;
    while (ctxLen < 100 && ctxStart > 0) {
      ctxStart--;
      ctxLen++;
      if (source[ctxStart] == "\n") {
        if (++ctxLines == 3) {
          break;
        }
      }
    }
    ctxLen = 0;
    ctxLines = 0;
    while (ctxLen < 100 && ctxEnd < source.length - 1) {
      ctxEnd++;
      ctxLen++;
      if (source[ctxEnd] == "\n") {
        if (++ctxLines == 3) {
          break;
        }
      }
    }
    var context = source.substring(ctxStart, this.location.offset) +
        "[ERROR ->]" +
        source.substring(this.location.offset, ctxEnd + 1);
    return '''${ this . msg} ("${ context}"): ${ this . location}''';
  }
}

class ParseSourceSpan {
  ParseLocation start;
  ParseLocation end;
  ParseSourceSpan(this.start, this.end) {}
  String toString() {
    return this
        .start
        .file
        .content
        .substring(this.start.offset, this.end.offset);
  }
}

File: es6/prod/src/compiler/parse_util.js
export class ParseLocation {
    constructor(file, offset, line, col) {
        this.file = file;
        this.offset = offset;
        this.line = line;
        this.col = col;
    }
    toString() { return `${this.file.url}@${this.line}:${this.col}`; }
}
export class ParseSourceFile {
    constructor(content, url) {
        this.content = content;
        this.url = url;
    }
}
export class ParseError {
    constructor(location, msg) {
        this.location = location;
        this.msg = msg;
    }
    toString() {
        var source = this.location.file.content;
        var ctxStart = this.location.offset;
        if (ctxStart > source.length - 1) {
            ctxStart = source.length - 1;
        }
        var ctxEnd = ctxStart;
        var ctxLen = 0;
        var ctxLines = 0;
        while (ctxLen < 100 && ctxStart > 0) {
            ctxStart--;
            ctxLen++;
            if (source[ctxStart] == "\n") {
                if (++ctxLines == 3) {
                    break;
                }
            }
        }
        ctxLen = 0;
        ctxLines = 0;
        while (ctxLen < 100 && ctxEnd < source.length - 1) {
            ctxEnd++;
            ctxLen++;
            if (source[ctxEnd] == "\n") {
                if (++ctxLines == 3) {
                    break;
                }
            }
        }
        let context = source.substring(ctxStart, this.location.offset) + '[ERROR ->]' +
            source.substring(this.location.offset, ctxEnd + 1);
        return `${this.msg} ("${context}"): ${this.location}`;
    }
}
export class ParseSourceSpan {
    constructor(start, end) {
        this.start = start;
        this.end = end;
    }
    toString() {
        return this.start.file.content.substring(this.start.offset, this.end.offset);
    }
}
